plugins/kubernetes/app/models/kubernetes/job_executor.rb

```
@@ -0,0 +1,78 @@
+# Execute a repository setup job
+# TODO unify/refactor with JobExecution ???
+module Kubernetes
+  class JobExecutor
+    WAIT_FOR_JOB = 5.minutes
+    TICK = 2.seconds
+
+    attr_reader :job
+    delegate :id, to: :job
+
+    def initialize(output, job:)
+      @output = output
+      @job = job
+    end
+
+    def default_build_cluster
+      Kubernetes::Cluster.first!
+    end
+
+    def client
+      default_build_cluster.client
+    end
+
+    def batch_client
+      default_build_cluster.batch_client
+    end
+
+    def build_job_config_path
+      File.join(Rails.root, 'plugins', 'kubernetes', 'app', 'models', 'build_job.yml')
+    end
+
+    def wait_for_job_status(kuber_job)
+      start = Time.now
+      loop do
+        job = Kubernetes::Api::Job.new(
+          batch_client.get_job(kuber_job.metadata.name, kuber_job.metadata.namespace))
+
+        return false, nil if job.failure?
+        return false, nil if start + WAIT_FOR_JOB < Time.now
+        return true, job if job.complete?
+
+        sleep TICK
+      end
+    end
+
+    def build!(reference, repo:, tag:, push: false, registry:)
+      @output.puts "### Running a remote Docker build job"
+      File.open(build_job_config_path, 'r') do |contents|
+        kuber_job = Kubernetes::JobConfigFile.new(contents, build_job_config_path).job
+        kuber_job.first_container.args = [@job.project.repository_directory, reference, repo, tag, push ? "yes" : "no"]
+        kuber_job.first_container.env = [
+          {name: 'DOCKER_REGISTRY_USER', value: registry["username"] || ""},
```

maybe fail if any of these are missing ?